expression/builtin_convert_charset.go

```diff
@@ -108,28 +127,155 @@ func (b *builtinInternalToBinarySig) vecEvalString(input *chunk.Chunk, result *c
+	if b.tp.Charset == charset.CharsetUTF8MB4 || b.tp.Charset == charset.CharsetUTF8 {
+		transferString = func(s string) (string, error) {
+			if !utf8.ValidString(s) {
+				return "", errCannotConvertString.GenWithStackByArgs(fmt.Sprintf("%X", s), charset.CharsetBin, b.tp.Charset)
 			}
-			sf := &ScalarFunction{
-				FuncName: model.NewCIStr(InternalFuncToBinary),
-				RetType:  exprTp,
-				Function: sig,
+			return s, nil
+		}
+	} else {
+		enc := charset.NewEncoding(b.tp.Charset)
+		transferString = func(s string) (string, error) {
+			str, err := enc.DecodeString(s)
+			if err != nil {
+				return "", errCannotConvertString.GenWithStackByArgs(fmt.Sprintf("%X", s), charset.CharsetBin, b.tp.Charset)
 			}
-			return FoldConstant(sf)
+			return str, nil
+		}
+	}
+	return transferString
+}
+
+// BuildToBinaryFunction to_binary function.
+func BuildToBinaryFunction(ctx sessionctx.Context, expr Expression) (res Expression) {
+	fc := &tidbToBinaryFunctionClass{baseFunctionClass{InternalFuncToBinary, 1, 1}}
+	f, err := fc.getFunction(ctx, []Expression{expr})
+	if err != nil {
+		return expr
+	}
+	res = &ScalarFunction{
+		FuncName: model.NewCIStr(InternalFuncToBinary),
+		RetType:  f.getRetTp(),
+		Function: f,
+	}
+	return FoldConstant(res)
+}
+
+// BuildFromBinaryFunction build from_binary function.
+func BuildFromBinaryFunction(ctx sessionctx.Context, expr Expression, tp *types.FieldType) (res Expression) {
+	fc := &tidbFromBinaryFunctionClass{baseFunctionClass{InternalFuncFromBinary, 1, 1}, tp}
+	f, err := fc.getFunction(ctx, []Expression{expr})
+	if err != nil {
+		return expr
+	}
+	res = &ScalarFunction{
+		FuncName: model.NewCIStr(InternalFuncFromBinary),
+		RetType:  tp,
+		Function: f,
+	}
+	return FoldConstant(res)
+}
+
+// HandleBinaryLiteral wraps `expr` with to_binary or from_binary sig.
+func HandleBinaryLiteral(ctx sessionctx.Context, expr Expression, ec *ExprCollation, funcName string) Expression {
+	switch funcName {
+	case ast.Concat, ast.ConcatWS, ast.Lower, ast.Lcase, ast.Reverse, ast.Upper, ast.Ucase, ast.Quote, ast.Coalesce,
+		ast.Left, ast.Right, ast.Repeat, ast.Trim, ast.LTrim, ast.RTrim, ast.Substr, ast.SubstringIndex, ast.Replace,
+		ast.Substring, ast.Mid, ast.Translate, ast.InsertFunc, ast.Lpad, ast.Rpad, ast.Elt, ast.ExportSet, ast.MakeSet,
+		ast.FindInSet, ast.Regexp, ast.Field, ast.Locate, ast.Instr, ast.Position, ast.GE, ast.LE, ast.GT, ast.LT, ast.EQ,
+		ast.NE, ast.NullEQ, ast.Strcmp, ast.If, ast.Ifnull, ast.Like, ast.In, ast.DateFormat, ast.TimeFormat:
+		if ec.Charset == charset.CharsetBin && expr.GetType().Charset != charset.CharsetBin {
+			return BuildToBinaryFunction(ctx, expr)
+		} else if ec.Charset != charset.CharsetBin && expr.GetType().Charset == charset.CharsetBin {
+			ft := expr.GetType().Clone()
+			ft.Charset, ft.Collate = ec.Charset, ec.Collation
+			return BuildFromBinaryFunction(ctx, expr, ft)
+		}
+	case ast.Hex, ast.Length, ast.OctetLength, ast.ASCII, ast.ToBase64:
```

Comment:
We can add `ast.MD5`, `ast.SHA1` and else here, then delete related code in expression.

Reply:
yes, and it will be in another pr

Reply:
Found it, #29950